### PR TITLE
Test that `dhall format` is idempotent

### LINKED ISSUE #1421: Test that `dhall format` is idempotent
As I mentioned in #1413, `dhall format` should be idempotent. That is, given some expression `x`, we want the outputs of `echo x | dhall format` and `echo x | dhall format | dhall format` to be the same.

This can be tested with a property test. [The existing `Src` generator](https://github.com/dhall-lang/dhall-haskell/blob/3f6a5ad8dfea4c1fa46c34fdd7d7f7dde2bcb319/dhall/tests/Dhall/Test/QuickCheck.hs#L303-L306) could be updated to produce valid and interesting comments. Source positions are AFAIK irrelevant.

### PR DESCRIPTION
**Pull request by basile-henry**
Closes #1421 

I am not too sure what kind of expression we expect `dhall format` to handle. Should `dhall format` be able to format code that doesn't type check?

Issue #1421 suggested using the `Arbitrary` instance for `Src` (which in turn uses the `Arbitrary` instance for `Text`) but this resulted in too many parse error we have to discard. I am now using the `Arbitrary` instance for `Expr s a` which seem much more reasonable.

This PR currently doesn't pass this new formatting test. However I am still opening it in order to discuss whether the test is too strict and to find out if anyone has any ideas for how to solve these remaining idempotency issues.

### CODE REVIEW

**Comment by sjakobi** [comment]
Thanks for working on this! :)

Looks like this is trickier than I had initially assumed! While it seems desirable that `dhall format` be idempotent for any input we throw at it, it would already be progress if we'd have idempotency for well-typed inputs.

> Issue #1421 suggested using the `Arbitrary` instance for `Src` (which in turn uses the `Arbitrary` instance for `Text`) but this resulted in too many parse error we have to discard. I am now using the `Arbitrary` instance for `Expr s a` which seem much more reasonable.

Ah right, I should have been clearer. What seemed to be tricky to get right is idempotency when we have comments in the source code. `Let`-`Binding`s can contain comments in several `Src` fields:

https://github.com/dhall-lang/dhall-haskell/blob/a343b19a632e3a6f964c80d33ad9f8317f04865e/dhall/src/Dhall/Core.hs#L2459-L2479

To test this, we could modify the `Src` generator to produce valid comments.

**Comment by basile-henry** [comment]
I have identified an issue with the header generation in my tests. I think I know how to fix it and will do so as soon as I can! :) I haven't abandoned this PR, just need a bit more time.

**Comment by joneshf** [comment]
Are there two issues that are trying to be solved? It seems like validating that `dhall format` is idempotent and validating that `dhall format` doesn't do weird things with comments are separate (the former subsuming the latter). Maybe testing that `./tests/format/*B.dhall` is equal to itself is enough? I.e. instead of assuming we'll have an `*A.dhall` and a `*B.dhall`: https://github.com/dhall-lang/dhall-haskell/blob/c284c19c116ce13947ba679be87ab5e87e2b4be0/dhall/tests/Dhall/Test/Format.hs#L49-L50
abstract over the input and output files so we test a `*B.dhall` against a `*B.dhall`.

**Comment by sjakobi** [comment]
> It seems like validating that `dhall format` is idempotent and validating that `dhall format` doesn't do weird things with comments are separate (the former subsuming the latter).

I wouldn't mind if these things were addressed incrementally. Initially we could work without comments, but add them later on.

> Maybe testing that `./tests/format/*B.dhall` is equal to itself is enough?

We can do that, but IMHO it would be a lot of work to get the coverage that a property test can provide, and it would be difficult to keep the coverage up-to-date.

**Comment by basile-henry** [comment]
Right, so it looks like the `Arbitrary` instances for `Expr` don't generate valid expressions. The expressions generated can't be parsed after being formatted. I don't mind fixing this but I think it'll be a bit more involved than I first thought.

Before I continue, I'd like to have your point of view on some implementation details. I started validating some `Arbitrary` instances by using `suchThat` and then checking the generated token is parsable. While this is nice conceptually it's likely terrible from a performance point of view.

I am thinking a better approach would be to create a module that contains generators for all the tokens that can be found in `Dhall.Parser.Token` and then build up from that to create the `Arbitrary` instances. Would you agree this is better? Any other ideas? :smile:

**Comment by sjakobi** [comment]
Sorry for underestimating the complexity of this issue!

Do you know what prevents the generated `Expr`s from being parsable? I see that you've started to address empty vars and labels – are there more big problems?

IMHO it's not necessary that every generated `Expr` is parsable – we can filter these out after the first formatting-parsing step. This is what we already do with the existing property tests for well-typed expressions. Of course the ratio of parsable expressions shouldn't be too low – but something like 20% is OK IMHO. We can simply run the property tests manually with a huge number of generated test cases when relevant changes come up. I've done this before to catch tricky disparencies between `isNormalized` and `normalize`.

> I am thinking a better approach would be to create a module that contains generators for all the tokens that can be found in `Dhall.Parser.Token` and then build up from that to create the `Arbitrary` instances. Would you agree this is better?

That might be viable but I believe that it will be tricky to maintain. Until a few weeks ago the `Expr` generator was hand-coded and was missing a bunch of `Expr` constructors – moving to `generic-random` fixed this, but this seems be more difficult with parser tokens.

I believe that improving the `Expr` generator could also pay off for the existing property tests, for example for pretty-printing test failures which is currently prevented by empty `Var`s and labels.

**Comment by basile-henry** [comment]
> Do you know what prevents the generated Exprs from being parsable? I see that you've started to address empty vars and labels – are there more big problems?

I think pretty much every single `Text` in the `Expr` AST has some invariants that are checked by the parser. And these are not checked/followed in the `Arbitrary` instances. Looking at the `dhall.abnf` file, it looks like there are quite a few rules about what characters are allowed in various places.

Right now, for example, I'm getting issues in the header comment because I'm using invalid characters:

```
Dhall Tests
  format tests
    Formatting should be idempotent: FAIL
      *** Failed! (after 5 tests):
      Exception:

        Error: Invalid input

        1:6:
          |
        1 | {-~F3
          |      ^^
        unexpected "<escape>-"
        expecting "-}", "{-", or crlf newline

        CallStack (from HasCallStack):
          error, called at tests/Dhall/Test/Format.hs:90:25 in main:Dhall.Test.Format
      Unicode
      (Header "{-~F3\ESC-}\n{-\b\844849%\732027-}\n{-\NUL\NAKY-}\n",Const Sort)
      Use --quickcheck-replay=801590 to reproduce.
```

> IMHO it's not necessary that every generated Expr is parsable – we can filter these out after the first formatting-parsing step. This is what we already do with the existing property tests for well-typed expressions.

I didn't realise we were doing this, if it's an acceptable strategy then I can revert the last 3 commits and try to modify the `maxDiscardRatio` setting. This is the kind of error I get if I revert them:

```
Dhall Tests
  format tests
    Formatting should be idempotent: FAIL (0.58s)
      *** Gave up! Passed only 48 tests; 1000 discarded tests.
      Use --quickcheck-replay=881203 to reproduce.
```

> I believe that improving the Expr generator could also pay off for the existing property tests

Generating parsable `Expr` would definitely speed up the tests but maybe that's outside the scope of the PR :sweat_smile:

File: dhall/src/Dhall/Util.hs [hunk 1]
```diff
@@ -12,6 +12,7 @@ module Dhall.Util
     , Output(..)
     , getExpression
     , getExpressionAndHeader
+    , Header(..)
```
Comment (sjakobi) [inline dhall/src/Dhall/Util.hs]: What's the motivation for this re-export? I'm wondering whether `Header` should live in `Dhall.Syntax`… :thinking:

File: dhall/src/Dhall/Parser.hs [hunk 2]
```diff
@@ -108,10 +121,10 @@ exprAndHeaderFromText
     :: String -- ^ User-friendly name describing the input expression,
               --   used in parsing error messages
     -> Text   -- ^ Input expression to parse
-    -> Either ParseError (Text, Expr Src Import)
+    -> Either ParseError (Header, Expr Src Import)
 exprAndHeaderFromText delta text = case result of
     Left errInfo   -> Left (ParseError { unwrap = errInfo, input = text })
-    Right (txt, r) -> Right (stripHeader txt, r)
+    Right (txt, r) -> Right (createHeader txt, r)
```
Comment (sjakobi) [inline dhall/src/Dhall/Parser.hs]: IMHO it would be better to leave the whitespace stripping to the pretty-printer. But that can be changed later on.

File: dhall/tests/Dhall/Test/Format.hs [hunk 3]
```diff
@@ -69,3 +80,13 @@ formatTest characterSet prefix =
                 <> "Actual   (show): " <> show actualText <> "\n"
 
         Tasty.HUnit.assertBool message (actualText == expectedText)
+
+idempotenceTest :: TestTree
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/Format.hs]: I think I'd slightly prefer to keep the property tests together in `D.T.QuickCheck`.

File: dhall/src/Dhall/Util.hs (hunk 1 above)
Comment (basile-henry) [reply dhall/src/Dhall/Util.hs]: I didn't mean to commit that. 😅 I used to have the `Header` newtype defined in that module. I'll remove the re-export!

File: dhall/src/Dhall/Parser.hs (hunk 2 above)
Comment (basile-henry) [reply dhall/src/Dhall/Parser.hs]: That would probably make more sense, I just left the whitespace stripping in the parser as it was. I'm also not too sure why it's done like this.

I think the concept of a header comment might make sense if we want to exploit it for documentation (maybe with markdown) in the future, but it needs to be a lot more strict. Currently it allows for multiple comments at once mixing single line and multiline.

File: dhall/src/Dhall/Util.hs (hunk 1 above)
Comment (basile-henry) [reply dhall/src/Dhall/Util.hs]: Ah wait no, I was mixing up with a different module. I re-exported it here, because several module use `Dhall.Util` to use `getExpressionAndHeader`, so it made sense to re-export `Header` with it.

What's the rational for what goes into `Dhall.Syntax`?

Comment (sjakobi) [reply dhall/src/Dhall/Util.hs]: > What's the rationale for what goes into `Dhall.Syntax`?

I wrote

https://github.com/dhall-lang/dhall-haskell/blob/ad443cd6851af215111e21d8ac92a9fb3cda7403/dhall/src/Dhall/Syntax.hs#L14-L19

But maybe `Headers` isn't a "core syntax type".

Let's leave things as they are now then. :+1:

File: dhall/tests/Dhall/Test/Format.hs (hunk 3 above)
Comment (basile-henry) [reply dhall/tests/Dhall/Test/Format.hs]: Moved

**Comment by basile-henry** [comment]
I have moved the test and set the number of `QuickCheckTests` to `10000` like we do in other property based tests. It looks like it started showing some actual issues with non-idempotent tests, for example:

```
Dhall Tests
  QuickCheck
    Formatting should be idempotent: FAIL (45.20s)
      *** Failed! Falsifiable (after 600 tests and 2 shrinks):
      Unicode
      (Header "",RecordCompletion (Embed (Import {importHashed = ImportHashed {hash = Nothing, importType = Missing}, importMode = Code})) (NaturalLit 0))
      "(missing)::0\n" /= "missing::0\n"
      Use --quickcheck-replay=970605 to reproduce.
```

**Comment by Gabriel439** [comment]
@basile-henry: So I dug into that example and it looks like the reason for non-idempotence is because the second time it formats the code it parses `missing` as a variable name rather than the keyword:

```
>>> fmap denote (exprFromText "" "(missing)::0")
Right (RecordCompletion (Var (V "missing" 0)) (NaturalLit 0))
```

**Comment by sjakobi** [comment]
Seems like the new test has found its first proper bug then! :)

```
$ echo '(missing)' | dhall

Use "dhall --explain" for detailed errors

Error: Unbound variable: missing

1│  missing

(stdin):1:2
```

On AppVeyor there's another failure that looks "real":

```
      *** Failed! Falsifiable (after 24 tests and 2 shrinks):
      ASCII
      (Header "{--}\n",Embed (Import {importHashed = ImportHashed {hash = Nothing, importType = Env "a?a"}, importMode = Code}))
      "{--}\nenv:a?a\n" /= "{--}\nenv:a ? a\n"
      Use --quickcheck-replay=980286 to reproduce.
  Input
```

How about we set the number of tests to 0 or 1 then to make sure they aren't flaky and merge. And then we can start triaging and fixing the bugs together?

**Comment by basile-henry** [comment]
> How about we set the number of tests to 0 or 1 then to make sure they aren't flaky and merge. And then we can start triaging and fixing the bugs together?

Sounds great! :+1:

**Comment by sjakobi** [comment]
One AppVeyor job failed with

```
    Formatting should be idempotent:                                                                                           FAIL
      *** Gave up! Passed only 0 tests; 10 discarded tests.
      Use --quickcheck-replay=11763 --quickcheck-max-size=10000 to reproduce.
  Input
```

File: dhall/tests/Dhall/Test/QuickCheck.hs [hunk 4]
```diff
@@ -504,13 +542,24 @@ tests =
         , embedThenExtractIsIdentity (Proxy :: Proxy (Vector Double))
         , embedThenExtractIsIdentity (Proxy :: Proxy (Data.Map.Map Double Bool))
         , embedThenExtractIsIdentity (Proxy :: Proxy (HashMap.HashMap Double Bool))
+        , ( "Formatting should be idempotent"
+          , idempotenceTest
+          , Test.Tasty.adjustOption (const $ QuickCheckTests 1) -- TODO Increase this!
+          . adjustQuickCheckMaxSize 10000 -- This test discards many cases
```
Comment (sjakobi) [inline dhall/tests/Dhall/Test/QuickCheck.hs]: How does the max-size adjustment relate to the comment?

Comment (basile-henry) [reply dhall/tests/Dhall/Test/QuickCheck.hs]: Ah :sweat_smile: I meant to use `QuickCheckMaxRatio`! It's a bit of a weird name, but as far as I understand it corresponds to the number of allowed discarded tests.

**Comment by sjakobi** [comment]
Cheers! :)

**Comment by sjakobi** [comment]
There's no way to run the test for a reasonable number of times without modifying the source right?

```
$ cabal test dhall:tasty --test-show-details=streaming --test-options "-p idempot --quickcheck-tests 10000000"
Build profile: -w ghc-8.6.5 -O0
In order, the following will be built (use -v for more details):
 - dhall-1.27.0 (lib) (configuration changed)
 - dhall-1.27.0 (test:tasty) (configuration changed)
Configuring library for dhall-1.27.0..
Preprocessing library for dhall-1.27.0..
Building library for dhall-1.27.0..
Configuring test suite 'tasty' for dhall-1.27.0..
Warning: The package has an extraneous version range for a dependency on an
internal library: dhall -any && ==1.27.0, dhall -any && ==1.27.0, dhall -any
&& ==1.27.0, dhall -any && ==1.27.0. This version range includes the current
package but isn't needed as the current package's library will always be used.
Preprocessing test suite 'tasty' for dhall-1.27.0..
Building test suite 'tasty' for dhall-1.27.0..
Running 1 test suites...
Test suite tasty: RUNNING...
Dhall Tests
  QuickCheck
    Formatting should be idempotent: OK
      +++ OK, passed 1 test; 1 discarded.
```

**Comment by basile-henry** [comment]
@sjakobi The issue is that the way this `QuickCheckTests` option was overridden was by using `max` to combine them: https://github.com/dhall-lang/dhall-haskell/blob/69b228454e4d96cd92de7f8f2b1abb7ebfb02e59/dhall/tests/Dhall/Test/QuickCheck.hs#L560

This means, it can only be overridden with a higher number. In the case of the new test, in order to set it to `1` I couldn't actually use the same adjust function (the default being 100): https://github.com/dhall-lang/dhall-haskell/blob/69b228454e4d96cd92de7f8f2b1abb7ebfb02e59/dhall/tests/Dhall/Test/QuickCheck.hs#L547

**Comment by sjakobi** [comment]
Maybe we can adjust with

    (`div` 100)

If you pass a really large `--quickcheck-tests`, we should then generate a good number of test cases.

**Comment by sjakobi** [comment]
Even better:

    subtract 99

;)
